File: packages/ai-parrot-server/ui/src/lib/components/agents/canvas/a2ui/linked/dsl.ts

function applyOp(rows: Row[], op: Op, frames: Record<string, Row[]>, key: string | null, i: number): Row[] {
  switch (op.op) {
    case 'select':
      return selectOp(rows, op, i);
    case 'rename':
      return renameOp(rows, op, i);
    case 'filter':
      return filterOp(rows, op, i);
    case 'sort':
      return sortOp(rows, op, i);
    case 'limit':
      return limitOp(rows, op, i);
    case 'derive':
      return deriveOp(rows, op, i);
    case 'group_by':
      return groupByOp(rows, op, frames, i);
    case 'pivot':
      return pivotOp(rows, op, i);
    case 'join':
      return joinOp(rows, op, frames, i);
    case 'union':
      return unionOp(rows, op, frames, i);
    default:
      throw new TransformError(`unknown op '${op.op}'`, key, i);
  }
}
function requireColumns(rows: Row[], columns: string[], opIndex: number, opName: string): void {
  if (rows.length === 0) return;
  const row = rows[0];
  const missing = columns.filter(column => !(column in row));
  if (missing.length > 0) {
    throw new TransformError(`${opName}: unknown column(s) ${JSON.stringify(missing)}; have ${JSON.stringify(Object.keys(row))}`, null, opIndex);
  }
}

function selectOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const columns = op.columns as string[];
  requireColumns(rows, columns, opIndex, 'select');
  return rows.map(row => {
    const newRow: Row = {};
    for (const col of columns) {
      newRow[col] = row[col];
    }
    return newRow;
  });
}

function renameOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const mapping = op.mapping as Record<string, string>;
  const fromColumns = Object.keys(mapping);
  requireColumns(rows, fromColumns, opIndex, 'rename');
  return rows.map(row => {
    const newRow: Row = {};
    for (const [key, value] of Object.entries(row)) {
      const newKey = key in mapping ? mapping[key] : key;
      newRow[newKey] = value;
    }
    return newRow;
  });
}

function filterOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const column = op.column as string;
  const operator = op.operator as string;
  const value = op.value;
  
  if (rows.length === 0) return [];
  requireColumns(rows, [column], opIndex, 'filter');
  
  return rows.filter(row => {
    const cellValue = row[column];
    // null never matches
    if (cellValue === null || cellValue === undefined) return false;
    
    switch (operator) {
      case 'eq': return cellValue === value;
      case 'ne': return cellValue !== value;
      case 'gt': return cellValue > value;
      case 'ge': return cellValue >= value;
      case 'lt': return cellValue < value;
      case 'le': return cellValue <= value;
      case 'in': 
        if (!Array.isArray(value)) {
          throw new TransformError("filter: 'in' requires a list value", null, opIndex);
        }
        return value.includes(cellValue);
      case 'contains':
        if (typeof cellValue === 'string' && typeof value === 'string') {
          return cellValue.includes(value);
        }
        return false;
      default:
        throw new TransformError(`filter: unsupported operator ${operator}`, null, opIndex);
    }
  });
}

function sortOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const by = op.by as Array<{column: string, direction: string}>;
  const columns = by.map(key => key.column);
  
  if (rows.length === 0) return [];
  requireColumns(rows, columns, opIndex, 'sort');
  
  // Make a copy to avoid mutating the original
  let sortedRows = [...rows];
  
  // Process in reverse order for stable sort
  for (let i = by.length - 1; i >= 0; i--) {
    const { column, direction } = by[i];
    sortedRows = sortedRows.sort((a, b) => {
      const aVal = a[column];
      const bVal = b[column];
      
      // nulls last
      if (aVal === null || aVal === undefined) return 1;
      if (bVal === null || bVal === undefined) return -1;
      
      // Handle date strings
      let aComparable = aVal;
      let bComparable = bVal;
      if (typeof aVal === 'string' && /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/.test(aVal)) {
        aComparable = Date.parse(aVal);
      }
      if (typeof bVal === 'string' && /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}/.test(bVal)) {
        bComparable = Date.parse(bVal);
      }
      
      if (direction === 'asc') {
        return aComparable < bComparable ? -1 : aComparable > bComparable ? 1 : 0;
      } else {
        return aComparable > bComparable ? -1 : aComparable < bComparable ? 1 : 0;
      }
    });
  }
  
  return sortedRows;
}

function limitOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const n = op.n as number;
  return rows.slice(0, n);
}

function deriveOperand(rows: Row[], expr: any, opIndex: number): any[] | number | string {
  if (typeof expr === 'string') {
    // Column reference
    requireColumns(rows, [expr], opIndex, 'derive');
    // Check if column is numeric
    if (rows.length > 0) {
      const sampleValue = rows[0][expr];
      if (typeof sampleValue !== 'number') {
        throw new TransformError(`derive: column ${JSON.stringify(expr)} is not numeric`, null, opIndex);
      }
    }
    return rows.map(row => row[expr]);
  }
  
  if (typeof expr === 'number') {
    return expr;
  }
  
  // Binary operation
  const left = deriveOperand(rows, expr.left, opIndex);
  const right = deriveOperand(rows, expr.right, opIndex);
  const operator = expr.operator;
  
  const leftArray = Array.isArray(left) ? left : rows.map(() => left);
  const rightArray = Array.isArray(right) ? right : rows.map(() => right);
  
  switch (operator) {
    case '+': return leftArray.map((l, i) => l + rightArray[i]);
    case '-': return leftArray.map((l, i) => l - rightArray[i]);
    case '*': return leftArray.map((l, i) => l * rightArray[i]);
    case '/': 
      return leftArray.map((l, i) => {
        const r = rightArray[i];
        if (r === 0) return null;
        return l / r;
      });
    default:
      throw new TransformError(`derive: unsupported operator ${operator}`, null, opIndex);
  }
}

function deriveOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const name = op.name as string;
  const expr = op.expr;
  
  const result = deriveOperand(rows, expr, opIndex);
  const resultArray = Array.isArray(result) ? result : rows.map(() => result);
  
  return rows.map((row, i) => ({
    ...row,
    [name]: resultArray[i]
  }));
}

function groupByOp(rows: Row[], op: Op, frames: Record<string, Row[]>, opIndex: number): Row[] {
  const by = op.by as string[];
  const aggregate = op.aggregate as Record<string, string>;
  
  requireColumns(rows, [...by, ...Object.keys(aggregate)], opIndex, 'group_by');
  
  if (rows.length === 0) return [];
  
  // Group rows by the 'by' columns
  const groups = new Map<string, Row[]>();
  for (const row of rows) {
    // Skip rows with null values in group columns
    if (by.some(col => row[col] === null || row[col] === undefined)) continue;
    
    const key = by.map(col => JSON.stringify(row[col])).join('|');
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key)!.push(row);
  }
  
  // Aggregate each group
  const result: Row[] = [];
  for (const [groupKey, groupRows] of groups) {
    const firstRow = groupRows[0];
    const newRow: Row = {};
    
    // Copy group by columns
    for (const col of by) {
      newRow[col] = firstRow[col];
    }
    
    // Apply aggregations
    for (const [col, aggFn] of Object.entries(aggregate)) {
      const values = groupRows.map(row => row[col]).filter(v => v !== null && v !== undefined);
      switch (aggFn) {
        case 'sum':
          newRow[col] = values.reduce((a, b) => (a as number) + (b as number), 0);
          break;
        case 'avg':
          newRow[col] = values.length > 0 ? values.reduce((a, b) => (a as number) + (b as number), 0) / values.length : null;
          break;
        case 'count':
          newRow[col] = values.length;
          break;
        case 'min':
          newRow[col] = values.length > 0 ? Math.min(...values as number[]) : null;
          break;
        case 'max':
          newRow[col] = values.length > 0 ? Math.max(...values as number[]) : null;
          break;
        default:
          throw new TransformError(`group_by: unsupported aggregate function ${aggFn}`, null, opIndex);
      }
    }
    
    result.push(newRow);
  }
  
  // Preserve first-appearance order
  return result;
}

function pivotOp(rows: Row[], op: Op, opIndex: number): Row[] {
  const indexCols = op.index as string[];
  const columnsCol = op.columns as string;
  const valuesCol = op.values as string;
  const aggregateFn = op.aggregate as string || 'sum';
  
  requireColumns(rows, [...indexCols, columnsCol, valuesCol], opIndex, 'pivot');
  
  if (rows.length === 0) return [];
  
  // Group by index columns
  const groups = new Map<string, Row[]>();
  for (const row of rows) {
    // Skip rows with null values in index or columns column
    if ([...indexCols, columnsCol].some(col => row[col] === null || row[col] === undefined)) continue;
    
    const key = indexCols.map(col => JSON.stringify(row[col])).join('|');
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key)!.push(row);
  }
  
  // Get unique column values (preserving first appearance order)
  const columnValues: string[] = [];
  const seenColumnValues = new Set<string>();
  for (const row of rows) {
    const colValue = row[columnsCol];
    if (colValue !== null && colValue !== undefined) {
      const strValue = String(colValue);
      if (!seenColumnValues.has(strValue)) {
        seenColumnValues.add(strValue);
        columnValues.push(strValue);
      }
    }
  }
  
  // Build result rows
  const result: Row[] = [];
  for (const [groupKey, groupRows] of groups) {
    const firstRow = groupRows[0];
    const newRow: Row = {};
    
    // Copy index columns
    for (const col of indexCols) {
      newRow[col] = firstRow[col];
    }
    
    // Create pivot columns
    const valueMap = new Map<string, number[]>();
    for (const row of groupRows) {
      const colValue = row[columnsCol];
      const value = row[valuesCol];
      if (colValue !== null && colValue !== undefined && value !== null && value !== undefined) {
        const strColValue = String(colValue);
        if (!valueMap.has(strColValue)) {
          valueMap.set(strColValue, []);
        }
        valueMap.get(strColValue)!.push(Number(value));
      }
    }
    
    // Apply aggregation and populate pivot columns
    for (const colValue of columnValues) {
      const values = valueMap.get(colValue) || [];
      let aggregatedValue: number | null = null;
      
      switch (aggregateFn) {
        case 'sum':
          aggregatedValue = values.reduce((a, b) => a + b, 0);
          break;
        case 'avg':
          aggregatedValue = values.length > 0 ? values.reduce((a, b) => a + b, 0) / values.length : null;
          break;
        case 'count':
          aggregatedValue = values.length;
          break;
        case 'min':
          aggregatedValue = values.length > 0 ? Math.min(...values) : null;
          break;
        case 'max':
          aggregatedValue = values.length > 0 ? Math.max(...values) : null;
          break;
      }
      
      newRow[colValue] = aggregatedValue;
    }
    
    result.push(newRow);
  }
  
  return result;
}

function joinOp(rows: Row[], op: Op, frames: Record<string, Row[]>, opIndex: number): Row[] {
  const withKey = op.with as string;
  const how = op.how as string || 'inner';
  const on = op.on as Array<{left: string, right: string}>;
  
  // Check if sibling frame exists
  if (!(withKey in frames)) {
    throw new TransformError(`join: sibling source ${JSON.stringify(withKey)} was not executed`, null, opIndex);
  }
  
  const rightRows = frames[withKey];
  const leftKeys = on.map(pair => pair.left);
  const rightKeys = on.map(pair => pair.right);
  
  requireColumns(rows, leftKeys, opIndex, 'join');
  requireColumns(rightRows, rightKeys, opIndex, 'join');
  
  // Identify column name collisions for prefixing
  const sameNamedKeys = new Set(rightKeys.filter((rightKey, i) => rightKey === leftKeys[i]));
  const rightOutputColumns = rightRows.length > 0 ? Object.keys(rightRows[0]).filter(col => !sameNamedKeys.has(col)) : [];
  const renamedColumns = new Map<string, string>();
  for (const col of rightOutputColumns) {
    if (rows.length > 0 && col in rows[0]) {
      renamedColumns.set(col, `${withKey}_${col}`);
    }
  }
  
  // Prepare right frame with renamed columns
  const rightWork = rightRows.map(row => {
    const newRow: Row = {};
    for (const [key, value] of Object.entries(row)) {
      const newKey = renamedColumns.get(key) || key;
      newRow[newKey] = value;
    }
    return newRow;
  });
  
  const mergeRightKeys = rightKeys.map(key => renamedColumns.get(key) || key);
  const outputRightColumns = rightOutputColumns.map(col => renamedColumns.get(col) || col);
  
  // Filter out rows with null keys
  const validLeftRows = rows.filter(row => leftKeys.every(key => row[key] !== null && row[key] !== undefined));
  const validRightRows = rightWork.filter(row => mergeRightKeys.every(key => row[key] !== null && row[key] !== undefined));
  
  // Perform join
  let result: Row[] = [];
  
  if (how === 'inner') {
    // Inner join
    for (const leftRow of validLeftRows) {
      const leftValues = leftKeys.map(key => leftRow[key]);
      
      for (const rightRow of validRightRows) {
        const rightValues = mergeRightKeys.map(key => rightRow[key]);
        
        // Check if keys match
        if (leftValues.every((val, i) => val === rightValues[i])) {
          const newRow: Row = { ...leftRow };
          for (const col of outputRightColumns) {
            newRow[col] = rightRow[col];
          }
          result.push(newRow);
        }
      }
    }
  } else if (how === 'left') {
    // Left join
    for (const leftRow of validLeftRows) {
      const leftValues = leftKeys.map(key => leftRow[key]);
      let matched = false;
      
      for (const rightRow of validRightRows) {
        const rightValues = mergeRightKeys.map(key => rightRow[key]);
        
        // Check if keys match
        if (leftValues.every((val, i) => val === rightValues[i])) {
          const newRow: Row = { ...leftRow };
          for (const col of outputRightColumns) {
            newRow[col] = rightRow[col];
          }
          result.push(newRow);
          matched = true;
        }
      }
      
      // If no match found, add left row with nulls for right columns
      if (!matched) {
        const newRow: Row = { ...leftRow };
        for (const col of outputRightColumns) {
          newRow[col] = null;
        }
        result.push(newRow);
      }
    }
    
    // Add rows with null keys from left
    const nullLeftRows = rows.filter(row => !leftKeys.every(key => row[key] !== null && row[key] !== undefined));
    for (const nullRow of nullLeftRows) {
      const newRow: Row = { ...nullRow };
      for (const col of outputRightColumns) {
        newRow[col] = null;
      }
      result.push(newRow);
    }
  }
  
  return result;
}

function unionOp(rows: Row[], op: Op, frames: Record<string, Row[]>, opIndex: number): Row[] {
  const sources = op.sources as string[];
  
  // Get all source frames
  const allFrames: Row[][] = [rows];
  for (const sourceKey of sources) {
    if (!(sourceKey in frames)) {
      throw new TransformError(`union: sibling source ${JSON.stringify(sourceKey)} was not executed`, null, opIndex);
    }
    allFrames.push(frames[sourceKey]);
  }
  
  if (allFrames.length === 0) return [];
  
  // Find column intersection
  let commonColumns: string[] = [];
  if (allFrames[0].length > 0) {
    commonColumns = Object.keys(allFrames[0][0]);
    for (let i = 1; i < allFrames.length; i++) {
      if (allFrames[i].length > 0) {
        const frameColumns = new Set(Object.keys(allFrames[i][0]));
        commonColumns = commonColumns.filter(col => frameColumns.has(col));
      }
    }
  }
  
  // Concatenate frames with common columns
  const result: Row[] = [];
  for (const frame of allFrames) {
    for (const row of frame) {
      const newRow: Row = {};
      for (const col of commonColumns) {
        newRow[col] = row[col];
      }
      result.push(newRow);
    }
  }
  
  return result;
}

File: packages/ai-parrot/src/parrot/outputs/a2ui/linked/dsl.py
"""Transform DSL v1 — Python reference executor (FEAT-598, spec §3 M2, §7).

``select`` keeps and orders columns; ``rename`` maps column names; ``filter`` supports
``eq|ne|gt|ge|lt|le|in|contains`` and never matches nulls; ``group_by`` aggregates;
``sort`` is stable with nulls last; ``limit`` truncates; and ``derive`` evaluates a
binary arithmetic tree. ``pivot``, ``join``, and ``union`` complete the closed DSL.
Rows use ``orient="records"``; dates serialise as ISO-8601; numeric dtypes are preserved;
and transform failures carry matching source key and operation index across executors.

Pure: no I/O; inputs are never mutated. ``ref`` transforms are renderer-side TypeScript and are
skipped here (the caller records ``transform_skipped: ref``).
"""

from __future__ import annotations

import json
import logging
from collections.abc import Callable, Mapping
from typing import TYPE_CHECKING, Any

from parrot.outputs.a2ui.linked.models import TransformSpec

if TYPE_CHECKING:
    import pandas as pd

logger = logging.getLogger(__name__)

__all__ = ["TransformError", "apply_transform", "frame_from_records", "frame_to_records"]


class TransformError(Exception):
    """A DSL op failed: missing column, derive type mismatch, absent join key, …

    Attributes:
        source_key: Linked-source key (dataModel root key), or ``None`` inside
            :func:`apply_transform` before the caller annotates it.
        op_index: 0-based index of the failing operation.
    """

    def __init__(self, source_key: str | None, op_index: int, message: str) -> None:
        """Initialise a transform failure with its execution location."""
        super().__init__(message)
        self.source_key = source_key
        self.op_index = op_index
        self.message = message


def frame_from_records(rows: list[dict[str, Any]]) -> "pd.DataFrame":
    """Build a frame from ``orient="records"`` rows (fixture/renderer shape)."""
    import pandas as pd

    return pd.DataFrame.from_records(rows)


def frame_to_records(frame: "pd.DataFrame") -> list[dict[str, Any]]:
    """Serialise records with ISO dates, JSON nulls, and preserved numeric values."""
    return json.loads(frame.to_json(orient="records", date_format="iso"))


_OpFn = Callable[["pd.DataFrame", Any, Mapping[str, "pd.DataFrame"], int], "pd.DataFrame"]
_DISPATCH: dict[str, _OpFn] = {}


def _register(op_name: str) -> Callable[[_OpFn], _OpFn]:
    """Register an operation implementation under its wire discriminator."""

    def deco(fn: _OpFn) -> _OpFn:
        _DISPATCH[op_name] = fn
        return fn

    return deco


def _require_columns(frame: "pd.DataFrame", columns: list[str], op_index: int, op_name: str) -> None:
    """Raise when an operation references columns not present in its input."""
    missing = [column for column in columns if column not in frame.columns]
    if missing:
        raise TransformError(None, op_index, f"{op_name}: unknown column(s) {missing}; have {list(frame.columns)}")


def apply_transform(
    frame: "pd.DataFrame", spec: TransformSpec | None, *, frames: Mapping[str, "pd.DataFrame"]
) -> "pd.DataFrame":
    """Apply ``spec.ops`` in order and return a new frame.

    A ``None`` transform copies the input; a renderer-side ``ref`` returns its
    original frame unchanged.
    """
    if spec is None:
        return frame.copy()
    if spec.ref is not None:
        return frame

    out = frame.copy()
    for index, op in enumerate(spec.ops or []):
        fn = _DISPATCH.get(op.op)
        if fn is None:
            raise TransformError(None, index, f"op {op.op!r} is not implemented")
        out = fn(out, op, frames, index)
    return out


@_register("select")
def _op_select(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Keep the requested columns in their requested order."""
    _require_columns(frame, list(op.columns), index, "select")
    return frame[list(op.columns)].copy()


@_register("rename")
def _op_rename(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Rename declared columns while retaining all columns' relative positions."""
    _require_columns(frame, list(op.mapping), index, "rename")
    return frame.rename(columns=dict(op.mapping)).copy()


@_register("filter")
def _op_filter(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Filter non-null values using one declared comparison operator."""
    if frame.empty and op.column not in frame.columns:
        return frame.copy()
    _require_columns(frame, [op.column], index, "filter")
    series = frame[op.column]
    valid = series.notna()
    if op.operator == "eq":
        matches = series == op.value
    elif op.operator == "ne":
        matches = series != op.value
    elif op.operator == "gt":
        matches = series > op.value
    elif op.operator == "ge":
        matches = series >= op.value
    elif op.operator == "lt":
        matches = series < op.value
    elif op.operator == "le":
        matches = series <= op.value
    elif op.operator == "in":
        if not isinstance(op.value, list):
            raise TransformError(None, index, "filter: 'in' requires a list value")
        matches = series.isin(op.value)
    elif op.operator == "contains":
        matches = series.map(lambda value: isinstance(value, str) and str(op.value) in value)
    else:
        raise TransformError(None, index, f"filter: unsupported operator {op.operator!r}")
    return frame.loc[valid & matches.fillna(False)].copy()


@_register("sort")
def _op_sort(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Perform a stable multi-key ordering with nulls always last."""
    columns = [key.column for key in op.by]
    if frame.empty and any(column not in frame.columns for column in columns):
        return frame.copy()
    _require_columns(frame, columns, index, "sort")
    out = frame.copy()
    for key in reversed(op.by):
        out = out.sort_values(
            by=key.column,
            ascending=key.direction == "asc",
            kind="mergesort",
            na_position="last",
        )
    return out


@_register("limit")
def _op_limit(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Return at most the declared number of leading rows."""
    return frame.head(int(op.n)).copy()


def _derive_operand(frame: "pd.DataFrame", expr: Any, index: int) -> Any:
    """Evaluate one numeric derive operand without interpreting arbitrary code."""
    import pandas as pd

    if isinstance(expr, str):
        _require_columns(frame, [expr], index, "derive")
        series = frame[expr]
        if not pd.api.types.is_numeric_dtype(series) or pd.api.types.is_bool_dtype(series):
            raise TransformError(None, index, f"derive: column {expr!r} is not numeric")
        return series
    if isinstance(expr, (int, float)) and not isinstance(expr, bool):
        return expr

    left = _derive_operand(frame, expr.left, index)
    right = _derive_operand(frame, expr.right, index)
    if expr.operator == "+":
        return left + right
    if expr.operator == "-":
        return left - right
    if expr.operator == "*":
        return left * right
    if expr.operator == "/":
        if isinstance(right, pd.Series):
            return left / right.mask(right == 0)
        if right == 0:
            if isinstance(left, pd.Series):
                return pd.Series(float("nan"), index=left.index)
            return None
        return left / right
    raise TransformError(None, index, f"derive: unsupported operator {expr.operator!r}")


@_register("derive")
def _op_derive(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Add a column calculated from a declarative numeric binary expression."""
    out = frame.copy()
    out[op.name] = _derive_operand(out, op.expr, index)
    return out


_AGGS = {"sum": "sum", "avg": "mean", "count": "count", "min": "min", "max": "max"}


def _sibling(frames: Mapping[str, "pd.DataFrame"], key: str, index: int, op_name: str) -> "pd.DataFrame":
    """Return an already-executed sibling frame or raise a location-aware error."""
    if key not in frames:
        raise TransformError(None, index, f"{op_name}: sibling source {key!r} was not executed")
    return frames[key]


@_register("group_by")
def _op_group_by(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Aggregate non-null groups while preserving their first-appearance order."""
    by = list(op.by)
    aggregate = dict(op.aggregate)
    _require_columns(frame, by + list(aggregate), index, "group_by")
    out = frame.groupby(by, sort=False, dropna=True).agg({column: _AGGS[fn] for column, fn in aggregate.items()})
    return out.reset_index()[by + list(aggregate)]


@_register("pivot")
def _op_pivot(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Pivot one value column into first-appearance ordered aggregate columns."""
    _require_columns(frame, list(op.index) + [op.columns, op.values], index, "pivot")
    out = frame.pivot_table(
        index=list(op.index),
        columns=op.columns,
        values=op.values,
        aggfunc=_AGGS[op.aggregate],
        sort=False,
    ).reset_index()
    return out.rename(columns={column: str(column) for column in out.columns})


@_register("join")
def _op_join(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Join a sibling frame without allowing null keys to match each other."""
    import pandas as pd

    right = _sibling(frames, op.with_, index, "join")
    left_keys = [pair.left for pair in op.on]
    right_keys = [pair.right for pair in op.on]
    _require_columns(frame, left_keys, index, "join")
    _require_columns(right, right_keys, index, "join")

    same_named_keys = {
        right_key for left_key, right_key in zip(left_keys, right_keys, strict=True) if left_key == right_key
    }
    right_output = [column for column in right.columns if column not in same_named_keys]
    renamed = {column: f"{op.with_}_{column}" for column in right_output if column in frame.columns}
    right_work = right.rename(columns=renamed).copy()
    merge_right_keys = [renamed.get(column, column) for column in right_keys]
    output_right = [renamed.get(column, column) for column in right_output]

    order_column = "__parrot_join_order"
    while order_column in frame.columns or order_column in right_work.columns:
        order_column = f"_{order_column}"
    left_work = frame.copy()
    left_work[order_column] = range(len(left_work))
    valid_left = left_work[left_work[left_keys].notna().all(axis=1)]
    valid_right = right_work[right_work[merge_right_keys].notna().all(axis=1)]
    merged = valid_left.merge(
        valid_right,
        how=op.how,
        left_on=left_keys,
        right_on=merge_right_keys,
        sort=False,
    )

    if op.how == "left":
        null_left = left_work[~left_work[left_keys].notna().all(axis=1)].copy()
        for column in output_right:
            null_left[column] = pd.NA
        merged = pd.concat([merged, null_left], ignore_index=True, sort=False)

    columns = list(frame.columns) + output_right
    return merged.sort_values(order_column, kind="mergesort")[columns].reset_index(drop=True)


@_register("union")
def _op_union(frame: "pd.DataFrame", op: Any, frames: Mapping[str, "pd.DataFrame"], index: int) -> "pd.DataFrame":
    """Concatenate the own and sibling frames on their ordered column intersection."""
    import pandas as pd

    others = [_sibling(frames, key, index, "union") for key in op.sources]
    columns = [column for column in frame.columns if all(column in other.columns for other in others)]
    return pd.concat([frame[columns], *(other[columns] for other in others)], ignore_index=True)
